File: backends/python/app/repositories/factory.py
from app.dbcontext import get_db_type, MONGO
from app.repositories.sql_repo import UserRepositorySQL, TaskRepositorySQL
from app.repositories.mongo_repo import UserRepositoryMongo, TaskRepositoryMongo

_user_sql = UserRepositorySQL()
_task_sql = TaskRepositorySQL()
_user_mongo = UserRepositoryMongo()
_task_mongo = TaskRepositoryMongo()


def get_user_repository():
    if get_db_type() == MONGO:
        return _user_mongo
    return _user_sql


def get_task_repository():
    if get_db_type() == MONGO:
        return _task_mongo
    return _task_sql

File: backends/node/src/repositories/factory.js
import { getDbType, MONGO } from '../dbcontext.js';
import { createUserRepositorySql, createUserRepositoryMongo } from './userRepository.js';
import { createTaskRepositorySql, createTaskRepositoryMongo } from './taskRepository.js';

export function createRepositoryFactory(db, mongoDb) {
  const userSql = createUserRepositorySql(db);
  const userMongo = mongoDb
    ? createUserRepositoryMongo(mongoDb.collection('users'))
    : null;
  const taskSql = createTaskRepositorySql(db);
  const taskMongo = mongoDb
    ? createTaskRepositoryMongo(mongoDb.collection('tasks'))
    : null;

  return {
    getUserRepository() {
      return getDbType() === MONGO && userMongo ? userMongo : userSql;
    },
    getTaskRepository() {
      return getDbType() === MONGO && taskMongo ? taskMongo : taskSql;
    },
  };
}
